keps/sig-node/24-apparmor-ga/README.md

```diff
@@ -0,0 +1,682 @@
+time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This is a proposal to upgrade the AppArmor annotation on pods to a dedicated
+field, and graduate the feature to GA. This proposal aims to do the _bare
+minimum_ to clean up the feature, without blocking future enhancements.
+
+## Motivation
+
+AppArmor support has been added with Kubernetes v1.4 and is already in beta.
+Profiles have to be available on each node whereas the container runtime ensures
+that the profile is loaded when specified at pod or PSP level. Profiles can be
+specified per-container via the pod’s metadata annotation:
+
+```
+container.apparmor.security.beta.kubernetes.io/<container_name>: {unconfined,runtime/default,localhost/<profile>}
+```
+
+The feature has been more or less unchanged ever since. Also note that the
+addition predates feature gates or our modern concept of feature lifecycle. So,
+even though the annotations include `beta` in the key, this is entirely useable
+on any production GA cluster.
+
+The main motivation behind this KEP is to promote the AppArmor feature to GA.
+
+_NOTE: Seccomp was in a very similar state, but with some subtle differences.
+Promoting Seccomp to GA will be covered by a [separate
+KEP](/keps/sig-node/20190717-seccomp-ga.md)._
+
+### Goals
+
+- Declare AppArmor as GA
+- Fully document and formally spec the feature support
+- Add equivalent API fields to replace AppArmor annotations
+- Deprecate the AppArmor annotations
+
+### Non-Goals
+
+This KEP proposes the absolute minimum to get AppArmor to GA, therefore all
+functional enhancements are out of scope, including:
+
+- Defining any standard "Kubernetes branded" AppArmor profiles
+- Formally specifying the AppArmor profile format in Kubernetes
+- Providing mechanisms for loading profiles from outside of the node
+- Changing the semantics around AppArmor support
+- Windows support
+
+## Proposal
+
+AppArmor is not available on every Linux distribution. Beside this, container
+runtimes have AppArmor as compile-time feature which may be disabled as well.
+With the GA API we do not change the error handling and behave exactly the same
+as the current error propagation paths.
+
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API. This
+includes the Pod API, which specifies what profile the containers run with.
+There are no changes planned to the current `PodSecurityPolicy` API as part of
+this KEP.
+
+#### Pod API
+
+The Pod AppArmor API is generally immutable, except in `PodTemplates`.
+
+```go
+type PodSecurityContext struct {
+    ...
+    // The AppArmor options to use by the containers in this pod.
+    // +optional
+    AppArmor  *AppArmorProfile
```

Comment:
pod-level apparmor profile is new to this KEP... the annotation-based feature only allows specifying container-level, correct?

Reply:
Yes that's correct. My idea was to apply the same hierarchy like we have it right now for seccomp. Basically if a profile gets added to a sandbox, it automatically applies to all containers. Containers security context have a higher priority than on the pod level. Do you think this is implementable? We do not pass the profile on sandbox creation via the CRI, so it would be just a pod to container field copy.

Reply:
I don't have enough context to know how tailored to a particular container AppArmor profiles typically are... is a pod-wide AppArmor profile likely to work with arbitrary injected sidecar containers?

Reply:
On the other hand, someone might prefer an injected sidecar picking up the pod-wide default and failing over running unconfined? Not sure...

Reply:
> I don't have enough context to know how tailored to a particular container AppArmor profiles typically are... is a pod-wide AppArmor profile likely to work with arbitrary injected sidecar containers?

At least the runtime/default profile (originated from docker) can be used with a wide variety of applications. From my viewpoint having a pod-wide field would be beneficial.

Reply:
I believe AppArmor's usage tend to be quite analogous to seccomp, although the former can be a lot more detailed, in both cases having profiles set at both Pod and Container levels would be quite useful, specially when considering multi-container pods.

For arbitrarily injected sidecars this becomes a bit of an interesting problem, as the pod level profile would need to account for its own workload plus an external workload that it would not know until deployment time. But on that case, the container-level profile override would resolve the issue, which is the same situation we have for seccomp.

Reply:
Yeah, I suspect with both AppArmor & seccomp the usecase for a pod-level profile is something akin to the default profile, with fairly broad grants. Application-specific profiles, will likely need to be tied to the container. With seccomp, the pod-level field seems slightly more useful since the "runtime default" profile is not actually the default, but I also agree that the APIs should probably mirror each other.